com.unity.multiplayer.mlapi/Runtime/Core/NetworkUpdateLoop.cs

```diff
@@ -234,131 +233,166 @@ public static PlayerLoopSystem CreateLoopSystem()
             }
         }
 
-        [RuntimeInitializeOnLoadMethod]
+        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.SubsystemRegistration)]
         private static void Initialize()
         {
-            var customPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+            UnregisterLoopSystems();
+            RegisterLoopSystems();
+        }
 
-            for (int i = 0; i < customPlayerLoop.subSystemList.Length; i++)
-            {
-                var playerLoopSystem = customPlayerLoop.subSystemList[i];
+        private enum LoopSystemPosition
+        {
+            After,
+            Before
+        }
 
-                if (playerLoopSystem.type == typeof(Initialization))
+        private static bool TryAddLoopSystem(ref PlayerLoopSystem parentLoopSystem, PlayerLoopSystem childLoopSystem, Type anchorSystemType, LoopSystemPosition loopSystemPosition)
+        {
+            int systemPosition = -1;
+            if (anchorSystemType != null)
+            {
+                for (int i = 0; i < parentLoopSystem.subSystemList.Length; i++)
                 {
-                    var subsystems = playerLoopSystem.subSystemList.ToList();
+                    var subsystem = parentLoopSystem.subSystemList[i];
+                    if (subsystem.type == anchorSystemType)
                     {
-                        // insert at the bottom of `Initialization`
-                        subsystems.Add(NetworkInitialization.CreateLoopSystem());
+                        systemPosition = loopSystemPosition == LoopSystemPosition.After ? i + 1 : i;
+                        break;
                     }
-                    playerLoopSystem.subSystemList = subsystems.ToArray();
                 }
-                else if (playerLoopSystem.type == typeof(EarlyUpdate))
+            }
+            else
+            {
+                systemPosition = loopSystemPosition == LoopSystemPosition.After ? parentLoopSystem.subSystemList.Length : 0;
+            }
+
+            if (systemPosition == -1) return false;
+
+            var newSubsystemList = new PlayerLoopSystem[parentLoopSystem.subSystemList.Length + 1];
+
+            // begin = systemsBefore + systemsAfter
+            // + systemsBefore
+            Array.Copy(parentLoopSystem.subSystemList, newSubsystemList, systemPosition);
+            // + childSystem
+            newSubsystemList[systemPosition] = childLoopSystem;
+            // + systemsAfter
+            Array.Copy(parentLoopSystem.subSystemList, systemPosition, newSubsystemList, systemPosition + 1, parentLoopSystem.subSystemList.Length - systemPosition);
```

Comment:
Just so I am clear is there a reason we are doing a copy twice? I feel like there is a missing check here.

Reply:
we copy from 0 to systemPosition (systemsBefore)
then we set childLoopSystem at systemPosition (childSystem)
and finally copy from systemPosition + 1 to the end (systemsAfter)

we don't double-copy, we half + 1 + half = all + 1 basically.
if the code doesn't read like that, I might comment more and/or introduce more local variables to indicate what's going on more clearly.